File: app/src/main/java/org/stypox/dicio/output/speech/SnackbarSpeechDevice.java
package org.stypox.dicio.output.speech;

import android.view.View;

import androidx.annotation.NonNull;

import com.google.android.material.snackbar.Snackbar;

public class SnackbarSpeechDevice extends InstantSpeechDevice {
    private View view;
    private Snackbar currentSnackbar = null;

    public SnackbarSpeechDevice(final View view) {
        this.view = view;
    }

    @Override
    public void speak(@NonNull final String speechOutput) {
        currentSnackbar = Snackbar.make(view, speechOutput, Snackbar.LENGTH_LONG);
        currentSnackbar.show();
    }

    @Override
    public void stopSpeaking() {
        if (currentSnackbar != null) {
            currentSnackbar.dismiss();
        }
    }

    @Override
    public void cleanup() {
        currentSnackbar = null;
        view = null;
    }
}

File: app/src/main/kotlin/org/stypox/dicio/io/speech/SnackbarSpeechDevice.kt
package org.stypox.dicio.io.speech

import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * This is always instantiated, but will do nothing if
 * it is not the speech device chosen by the user
 */
@Singleton
class SnackbarSpeechDevice @Inject constructor() : InstantSpeechDevice() {
    // null indicates that any snackbar should be dismissed
    private val _events = MutableSharedFlow<String?>(0, 1, BufferOverflow.DROP_OLDEST)
    val events: SharedFlow<String?> = _events

    override fun speak(speechOutput: String) {
        _events.tryEmit(speechOutput)
    }

    override fun stopSpeaking() {
        _events.tryEmit(null)
    }

    override fun cleanup() {
        // nothing to do
    }
}
